modules/clients/src/test/java/org/apache/ignite/jdbc/thin/JdbcThinBestEffortAffinityTransactionsSelfTest.java

```diff
@@ -0,0 +1,274 @@
+import org.apache.ignite.internal.processors.query.QueryHistoryMetrics;
+import org.apache.ignite.internal.processors.query.h2.IgniteH2Indexing;
+import org.apache.ignite.testframework.GridStringLogger;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Jdbc thin transactional best effort affinity test.
+ */
+public class JdbcThinBestEffortAffinityTransactionsSelfTest extends JdbcThinAbstractSelfTest {
+    /** */
+    private static final String URL = "jdbc:ignite:thin://127.0.0.1:10800..10802";
+
+    /** Nodes count. */
+    private static final int NODES_CNT = 3;
+
+    /** Query execution multiplier. */
+    private static final int QUERY_EXECUTION_MULTIPLIER = 5;
+
+    /** Connection. */
+    private Connection conn;
+
+    /** Statement. */
+    private Statement stmt;
+
+    /** {@inheritDoc} */
+    @SuppressWarnings({"deprecation", "unchecked"})
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        cfg.setCacheConfiguration(cacheConfiguration(DEFAULT_CACHE_NAME).setNearConfiguration(null));
+
+        cfg.setMarshaller(new BinaryMarshaller());
+
+        cfg.setGridLogger(log = new GridStringLogger());
+
+        return cfg;
+    }
+
+    /**
+     * @param name Cache name.
+     * @return Cache configuration.
+     */
+    private CacheConfiguration cacheConfiguration(@NotNull String name) {
+        CacheConfiguration cfg = defaultCacheConfiguration();
+
+        cfg.setName(name);
+        cfg.setAtomicityMode(CacheAtomicityMode.TRANSACTIONAL_SNAPSHOT);
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTestsStarted() throws Exception {
+        super.beforeTestsStarted();
+
+        startGridsMultiThreaded(NODES_CNT);
+
+        try (Connection c = prepareConnection(true, NestedTxMode.ERROR)) {
+            try (Statement stmt = c.createStatement()) {
+                stmt.execute("CREATE TABLE Person (id int primary key, firstName varchar, lastName varchar, age int) WITH \"cache_name=persons,wrap_value=true," +
+                    "atomicity=transactional_snapshot\"");
+
+                stmt.executeUpdate("insert into Person (id, firstName, lastName, age) values (1, 'John1', 'Dow1', 1);" +
+                    "insert into Person (id, firstName, lastName, age) values (2, 'John2', 'Dow2', 2);" +
+                    "insert into Person (id, firstName, lastName, age) values (3, 'John3', 'Dow3', 3);");
+            }
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        conn = DriverManager.getConnection(URL);
+
+        stmt = conn.createStatement();
+
+        assert stmt != null;
+        assert !stmt.isClosed();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        if (stmt != null && !stmt.isClosed()) {
```

Comment:
Simple U.closeQuiet(..) can be used here.

Reply:
fixed